docstr_coverage/result_collection.py

```diff
@@ -0,0 +1,312 @@
+"""Module containing the classes required to collect and aggregate docstring information.
+
+Currently, this module is in BETA and its interface may change in future versions."""
+
+import enum
+import functools
+import operator
+from typing import Optional
+
+
+class ResultCollection:
+    """A result collection contains information about the presence of docstrings collected during
+    the walk through the files. From the `ResultCollection`, information about missing docstrings
+    for inspected files can be retrieved, and summary metrics (e.g. coverage) can be extracted
+    without having to walk over the files again."""
+
+    def __init__(self):
+        self._files = dict()
+
+    def get_file(self, file_path):
+        """Provides access to the docstring information for specific files.
+        This primarily targets information collection phase
+        (thus the docstr-coverage internal process).
+
+        If no file with the provided name is on record yet, a new empty `File` information instance
+        is created and returned.
+
+        Parameters
+        ----------
+        file_path: String
+            The path of the file for which the information is requested.
+
+        Returns
+        -------
+        File
+            The file (information) instance used to track docstring information"""
+        try:
+            return self._files[file_path]
+        except KeyError:
+            file = File()
+            self._files[file_path] = file
+            return file
+
+    def count(self):
+        """Walks through all the tracked files in this result collection, and counts overall
+        statistics, such as #missing docstring.
+
+        Returns
+        -------
+        AggregatedCount
+            A count instance containing a range of docstring counts."""
+        counts = (folder.count() for folder in self._files.values())
+        return functools.reduce(operator.add, counts, AggregatedCount())
+
+    def files(self):
+        """Generator, iterating over all (file-name, file-info) tuples in this result collection"""
+        for file_path, file in self._files.items():
+            yield file_path, file
+
+    def to_legacy(self):
+        """Converts the information in this `ResultCollection` into the less expressive dictionary
+        of counts used since the early versions of docstr-coverage."""
+        file_results = dict()
+        for filename, file in self.files():
+            missing_list = [
+                e.node_identifier
+                for e in file.expected_docstrings()
+                if not e.ignore_reason
+                and not e.has_docstring
+                and not e.node_identifier == "module docstring"
+            ]
+            has_module_doc = (
```

Comment:
How do you feel about adding a boolean `has_module_docstring` property to the `File` class below, and referencing it here? We can probably default `has_module_docstring` to False, then set it to True inside `File.report_module`

Reply:
I see your motivation here, but I prefer my version for now.
The current version (string comparison) is a bit clumsy, but I chose it for three reasons:

1.  It's only temporary. I plan to add Docstr-Types (module, function, class, method) in the near future which would allow the creation of much nicer new printers, also summarizing coverage by docstr-type etc... This future change will allow to get rid of the string comparison.
2. The module docstring should for general consistency also be part of the iterator on all collected docstrings in a file. In general, I do not really like the special treatment of module docstrings (again motivated by some visions I have of future features). 

Now that I think about it, I should really create a roadmap issue stating my ideas for future major features (there are a few), allowing us to discuss them. :-)

Reply:
Or, do I misunderstand you and you suggest keeping the boolean in addition to the `ExpectedDocstring` in the collection? I think that could be a good trade-off here, actually...

Reply:
I see. That makes sense. 

> In general, I do not really like the special treatment of module docstrings

That's a good point. Better to treat them like any other docstring. It just happens to be placed at the top of a file.

> I should really create a roadmap issue stating my ideas for future major features (there are a few), allowing us to discuss them

That's an excellent idea!